src/TwitchLib.Communication/Clients/AClientBase.cs

```diff
@@ -0,0 +1,459 @@
+        }
+        #endregion invoker/raiser internal
+
+
+        #region methods public
+        public bool Send(string message)
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            try
+            {
+                lock (LOCK)
+                {
+                    SpecificClientSend(message);
+                    return true;
+                }
+            }
+            catch (Exception e)
+            {
+                RaiseSendFailed(new OnSendFailedEventArgs(message, e));
+                return false;
+            }
+        }
+
+        public bool Open()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            return OpenPrivate(false);
+        }
+        public void Close()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            // ConnectionWatchDog has to be stopped first
+            // so that it wont reconnect
+            NetworkServices.Stop();
+            // ClosePrivate() also handles IClientOptions.DisconnectWait
+            ClosePrivate();
+        }
+        /// <summary>
+        ///     <inheritdoc cref="Close"/>
+        /// </summary>
+        public void Dispose()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            Close();
+            GC.SuppressFinalize(this);
+        }
+
+        public bool Reconnect()
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            // stops everything, ConnectionWatchDog too
+            if (IsConnected)
+            {
+                Close();
+            }
+            // interface IClient doesnt declare a return value for Reconnect()
+            // so we can suppress IDE0058 of ReconnectInternal()
+            return ReconnectInternal();
+        }
+        #endregion methods public
+
+
+        #region methods private
+        private bool OpenPrivate(bool isReconnect)
+        {
+            LOGGER?.TraceMethodCall(GetType());
+            try
+            {
+                if (Token.IsCancellationRequested)
+                {
+                    return false;
+                }
+                if (IsConnected)
+                {
+                    return true;
+                }
+                // at this time,
+                // the specific 'System.Net.WebSockets.ClientWebSocket'
+                // or 'System.Net.Sockets.TcpClient'
+                // is null or it is disposed
+                // this has to be the only place where 'SetSpecificClient()' is called!!!
+                SetSpecificClient();
+                bool first = true;
+                Options.ReconnectionPolicy.Reset(isReconnect);
+                while (!IsConnected
+                       && !Options.ReconnectionPolicy.AreAttemptsComplete())
+                {
+                    LOGGER?.TraceAction(GetType(), "try to connect");
+                    if (!first)
+                    {
+                        Task.Delay(Options.ReconnectionPolicy.GetReconnectInterval()).GetAwaiter().GetResult();
```

Comment:
I was thinking about that as well.
I mean the whole thing is basically a breaking change already it has some API changes that propagate up to public user API space in TwitchLib.Client and TwitchLib.Pubsub anyway so we can go the full distance

Reply:
That's true. But I would go for second PR on this change as well. You don't have to merge into master and release immediately right? This can go in several PRs. This one, then async/await, then nullability change. It's always better to have several smaller PRs than one giant one.

Reply:
A commit/merge to master doesnt kick of a release.
It only targets master bcs I was an idiot myself and commited sth directly to master instead of dev and I didnt sync back up dev yet.
wanted to do that after this PR is merged.